File: mobile/src/lib/dice/notation.ts
/**
 * Parser de notação de dados — espelho de backend/app/services/dice/notation.py.
 * Usado para validar a entrada e para rolagens offline. Online, o servidor é quem rola.
 */

export const ALLOWED_SIDES = [2, 4, 5, 6, 8, 10, 12, 16, 20, 24, 30, 60, 100, 1000] as const;
export type DieSides = (typeof ALLOWED_SIDES)[number];

export const MAX_DICE = 100;
export const MAX_MODIFIER = 10_000;
export const MAX_NOTATION_LENGTH = 64;

export class NotationError extends Error {}

export type DiceTerm = {
  count: number;
  sides: DieSides;
  sign: 1 | -1;
  keep?: 'kh' | 'kl';
  keepN?: number;
};

export type DiceExpression = {
  terms: DiceTerm[];
  modifier: number;
};

const TERM = /(?:(\d*)d(\d+|%)(?:(kh|kl)(\d+))?|(\d+))/y;

export function isAllowedSides(n: number): n is DieSides {
  return (ALLOWED_SIDES as readonly number[]).includes(n);
}

export function parse(notation: string): DiceExpression {
  if (!notation || notation.length > MAX_NOTATION_LENGTH) {
    throw new NotationError('Notação vazia ou longa demais.');
  }
  const source = notation.replace(/\s+/g, '').toLowerCase();
  if (!source) throw new NotationError('Notação vazia.');

  const terms: DiceTerm[] = [];
  let modifier = 0;
  let pos = 0;
  let first = true;
  while (pos < source.length) {
    let sign: 1 | -1 = 1;
    const ch = source[pos];
    if (ch === '+' || ch === '-') {
      sign = ch === '-' ? -1 : 1;
      pos += 1;
    } else if (!first) {
      throw new NotationError(`Esperado '+' ou '-' na posição ${pos + 1}.`);
    }
    TERM.lastIndex = pos;
    const match = TERM.exec(source);
    if (!match || TERM.lastIndex === pos) {
      throw new NotationError(`Termo inválido na posição ${pos + 1}.`);
    }
    pos = TERM.lastIndex;
    first = false;

    const [, countRaw, sidesRaw, keep, keepRaw, constant] = match;
    if (constant !== undefined) {
      modifier += sign * Number(constant);
      continue;
    }
    const count = countRaw ? Number(countRaw) : 1;
    const sides = sidesRaw === '%' ? 100 : Number(sidesRaw);
    if (!isAllowedSides(sides)) throw new NotationError(`Dado d${sides} não é suportado.`);
    if (count < 1) throw new NotationError('A quantidade de dados deve ser pelo menos 1.');
    const keepN = keep ? Number(keepRaw) : undefined;
    if (keepN !== undefined && (keepN < 1 || keepN > count)) {
      throw new NotationError(`Não dá para manter ${keepN} de ${count} dados.`);
    }
    terms.push({ count, sides, sign, keep: keep as DiceTerm['keep'], keepN });
  }

  if (terms.length === 0) throw new NotationError('A expressão precisa ter pelo menos um dado.');
  if (terms.reduce((n, t) => n + t.count, 0) > MAX_DICE) {
    throw new NotationError(`No máximo ${MAX_DICE} dados por rolagem.`);
  }
  if (Math.abs(modifier) > MAX_MODIFIER) throw new NotationError('Modificador grande demais.');
  return { terms, modifier };
}

export function termToString(term: DiceTerm): string {
  return `${term.count}d${term.sides}${term.keep ? `${term.keep}${term.keepN}` : ''}`;
}

export function canonical(expr: DiceExpression): string {
  let out = '';
  expr.terms.forEach((term, i) => {
    if (term.sign < 0) out += '-';
    else if (i > 0) out += '+';
    out += termToString(term);
  });
  if (expr.modifier > 0) out += `+${expr.modifier}`;
  else if (expr.modifier < 0) out += String(expr.modifier);
  return out;
}

File: backend/app/services/dice/notation.py
"""Parser de notação de dados (sem eval).

Gramática aceita (espaços são ignorados, maiúsculas/minúsculas tanto faz):

    expressao := termo (('+' | '-') termo)*
    termo     := dados | inteiro
    dados     := [N] 'd' (LADOS | '%') ['!'] [('kh' | 'kl') K]

Exemplos: ``d20``, ``1d20+5``, ``2d20kh1`` (vantagem), ``4d6kh3``, ``d%``, ``3d6-2``, ``1d1000``.
``!`` = dado que explode (Savage Worlds): no valor máximo rola de novo e soma, ex.: ``1d8!``.
"""

import re
from dataclasses import dataclass

ALLOWED_SIDES = frozenset({2, 4, 5, 6, 8, 10, 12, 16, 20, 24, 30, 60, 100, 1000})
MAX_DICE = 100
MAX_MODIFIER = 10_000
MAX_NOTATION_LENGTH = 64

_TERM = re.compile(
    r"(?:(?P<count>\d*)d(?P<sides>\d+|%)(?P<explode>!)?(?:(?P<keep>kh|kl)(?P<keep_n>\d+))?|(?P<const>\d+))"
)


class NotationError(ValueError):
    pass


@dataclass(frozen=True)
class DiceTerm:
    count: int
    sides: int
    sign: int = 1
    keep: str | None = None  # "kh" ou "kl"
    keep_n: int | None = None
    explode: bool = False

    @property
    def kept_count(self) -> int:
        return self.keep_n if self.keep_n is not None else self.count

    def canonical(self) -> str:
        suffix = f"{self.keep}{self.keep_n}" if self.keep else ""
        return f"{self.count}d{self.sides}{'!' if self.explode else ''}{suffix}"


@dataclass(frozen=True)
class DiceExpression:
    terms: tuple[DiceTerm, ...]
    modifier: int

    def canonical(self) -> str:
        out = ""
        for i, term in enumerate(self.terms):
            if term.sign < 0:
                out += "-"
            elif i > 0:
                out += "+"
            out += term.canonical()
        if self.modifier > 0:
            out += f"+{self.modifier}"
        elif self.modifier < 0:
            out += str(self.modifier)
        return out


def parse(notation: str) -> DiceExpression:
    if not notation or len(notation) > MAX_NOTATION_LENGTH:
        raise NotationError("Notação vazia ou longa demais.")
    source = re.sub(r"\s+", "", notation).lower()
    if not source:
        raise NotationError("Notação vazia.")

    terms: list[DiceTerm] = []
    modifier = 0
    pos = 0
    first = True
    while pos < len(source):
        sign = 1
        if source[pos] in "+-":
            sign = -1 if source[pos] == "-" else 1
            pos += 1
        elif not first:
            raise NotationError(f"Esperado '+' ou '-' na posição {pos + 1}.")
        match = _TERM.match(source, pos)
        if not match or match.end() == pos:
            raise NotationError(f"Termo inválido na posição {pos + 1}.")
        pos = match.end()
        first = False

        if match.group("const") is not None:
            modifier += sign * int(match.group("const"))
            continue

        count = int(match.group("count")) if match.group("count") else 1
        sides = 100 if match.group("sides") == "%" else int(match.group("sides"))
        if sides not in ALLOWED_SIDES:
            raise NotationError(f"Dado d{sides} não é suportado.")
        if count < 1:
            raise NotationError("A quantidade de dados deve ser pelo menos 1.")
        keep = match.group("keep")
        keep_n = int(match.group("keep_n")) if keep else None
        if keep_n is not None and not 1 <= keep_n <= count:
            raise NotationError(f"Não dá para manter {keep_n} de {count} dados.")
        explode = match.group("explode") is not None
        if explode and sides < 3:
            raise NotationError("Só dados de 3 lados ou mais podem explodir.")
        terms.append(DiceTerm(count=count, sides=sides, sign=sign, keep=keep, keep_n=keep_n, explode=explode))

    if not terms:
        raise NotationError("A expressão precisa ter pelo menos um dado.")
    if sum(t.count for t in terms) > MAX_DICE:
        raise NotationError(f"No máximo {MAX_DICE} dados por rolagem.")
    if abs(modifier) > MAX_MODIFIER:
        raise NotationError("Modificador grande demais.")
    return DiceExpression(terms=tuple(terms), modifier=modifier)
